### ISSUE TITLE
Room for optimizing rc stop operations

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#8676]
We haven't really profiled how long it takes to stop a large rc and I suspect there's a lot of room for improvement. At the kubectl level we should do: https://github.com/GoogleCloudPlatform/kubernetes/issues/8572. At the rc manager level:
1. Resize rc to 0 -> "real time" (or as fast as the watch anyway)
2. Rc deletes pods -> rate limited to 20qps over all rcs (this connection is shared with node and endpoints controller), so however long it takes (probably as long as it takes to spin up the rc)
3. Update status.replicas -> "real time" (as the watch delivers notifications for each pod deleted)

Here is at least 1 corner case where updating status.Replicas can take upto 30s extra: 
when the pods controller does a relist (once in 5m) in between when the rc has fired off a bunch of deletes, these deletes have hit the apiserver, but before the watch has delivered those deletes to the rc manager. The relist will not have those pods but the store will. We do not differentiate this case from  dropped deletes: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L247 

One way to fix this would be for the informer to embed the deleted object instead of just the key, into the tombstone entry.

@kargakis @smarterclayton

**Comment by smarterclayton** [comment]
Agree resize to zero should always be as fast as possible within the rate limit of the RC client.  Freeing up cluster resources should be fast.

**Comment by kargakis** [comment]
+1, especially for 3. By playing around with rc and pod deletion so far, I can say it's a pain point.

**Comment by bprashanth** [comment]
1,2 and 3 are descriptions of what happens today. You are probably experiencing the rate limiting in 2. If you would like to experiment you can bump up the rate limit on your connection (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L189) and the BurstReplicas in the rc manager (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L69). Note that a high qps could have weird effects on your master components so I wouldn't go overboard.

**Comment by kargakis** [comment]
Not sure about it, because the pods were getting deleted fast w/o a problem (http://pastebin.com/gaLPpqSU). I observed that it takes a lot for rc.status.replicas to sync. The syncHandler of the rc manager from time to time needed to run more than once in order to update the rc with the desired replicas number (I am not familiar with the code but it seems to me that this shouldn't happen).

**Comment by bprashanth** [comment]
@kargakis in the pastebin I think the last pod dies at : 
```
I0519 13:32:09.041077   14299 replication_controller.go:331] Deleted pod "ruby-hello-world-1-s0lgz"
```

and the rc dies at
```
I0519 13:32:10.217783   14299 replication_controller.go:350] Replication Controller has been deleted test/ruby-hello-world-1
```

The status.Replicas should go down as each pod is dying unless something else is updating the rc concurrently (in which case it will take multiple sync calls). If you have many rcs, they share rc manager time, so if one fails for some reason it's requeued at the tail.

I'd be glad to help debug if you can elaborate on expected/observed.

**Comment by kargakis** [comment]
> I'd be glad to help debug if you can elaborate on expected/observed.

Sure thing, I will get back in testing for this later today or tomorrow.

**Comment by kargakis** [comment]
http://pastebin.com/jmb5JTYm

@bprashanth see for yourself. While resizng up to 20 pods is fast (>2s), updating the rc.status takes more than 4s.

**Comment by wojtek-t** [comment]
The corner case described in: https://github.com/GoogleCloudPlatform/kubernetes/issues/8676#issue-79510496 is something that we're hitting while running out scalability tests on Jenkins. And there are quite a lot of those causing test failures.

I also debugged it as part of #7561 - see: https://github.com/GoogleCloudPlatform/kubernetes/issues/7561#issuecomment-105512435

I think that fixing this should be done before v1.0 - cc @davidopp

**Comment by smarterclayton** [comment]
Agree, we're hitting it in scale reliability testing as well.

**Comment by bprashanth** [comment]
I'll get a fix today for the issue mentioned in my first comment.
Apart from that, there's also a heck of a lot of cpu contention because we fire off n goroutines to create/delete n replicas. There are levels of rate limiting that help, but in general turning down an rc is slower than starting one, because the latter is usually not competing for cpu/etcd time with anything else.  The solution to this is a batch create endpoint.

@kargakis still a little confused:

It fired off 20 goroutines to create pods (note that we can't set status.replicas at this point, since we've only created the goroutines to create pods. These still need to get cpu time):
```
I0526 13:18:42.952089    8112 replication_controller.go:302] Too few "test"/"ruby-hello-world-1" replicas, need 20, creating 19
```
-- at some point the goroutines ran, hit the apiserver, and the rc manager noticed the watch --

It updated status.Replicas:
```
I0526 13:18:44.873369    8112 controller_utils.go:306] Updating replica count for rc/ruby-hello-world-1, 1->20
```

**Comment by kargakis** [comment]
> -- at some point the goroutines ran, hit the apiserver, and the rc manager noticed the watch --
> 
> It updated status.Replicas:
> 
> I0526 13:18:44.873369    8112 controller_utils.go:306] Updating replica count for rc/ruby-hello-world-1, 1->20

[Here](https://github.com/kargakis/origin/blob/4f0dd02d4b63d33998441bc4d8c18b0076cff1ee/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/controller_utils.go#L306) it starts by making the UPDATE call though it gets updated much [later](https://github.com/kargakis/origin/blob/4f0dd02d4b63d33998441bc4d8c18b0076cff1ee/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/controller_utils.go#L314)
```
I0526 13:18:49.272060    8112 controller_utils.go:314] Successfully updated rc/ruby-hello-world-1
```

I can reproduce this pretty easy: http://pastebin.com/raw.php?i=2J9mcnhK (Cleaned-up kubelet and event logs)
One more thing to note is that the syncHandler misses one pod in the first pass
```
I0526 17:01:52.862598    8112 controller_utils.go:306] Updating replica count for rc/ruby-hello-world-1, 1->19
```

**Comment by bprashanth** [comment]
> One more thing to note is that the syncHandler misses one pod in the first pass

That depends on if the watch has transmitted the pod or not. Status.Replicas will not be be on the dot, there's bound to be some latency depending on the time it takes for an update to reflect in the watch. More specifically this is the rpc firing:
```
I0526 17:01:52.461230    8112 replication_controller.go:312] Created new pod
```
It will incur a network roundtrip for the watch to fire, after which the rc will wake up and update status.replicas. So taking multiple syncHandler calls is expected. 

The slow update is weird and warrants digging.

**Comment by saad-ali** [comment]
@davidopp @lavalamp Can you please triage this for pre/post 1.0. I will mark #8822 accordingly.

**Comment by bprashanth** [comment]
The impact is a longer timeout for `rc stop` (on avg it takes about 2.30 to tear down 3000 pods, can take upto 5m if it hits the mentioned case).

**Comment by bprashanth** [comment]
@kargakis re slow status.Replicas. Since I can't get it to repro on my setup with a single rc and 40 replicas I can try to debug your logs (as a data point sizing up/down one rc with 3000 replicas across 100 nodes, the status updates *sometimes* takes 2s on gce with an n1-standard-4 master).

You've refactored the loop a bit in your local branch in a way that doesn't enforce the retry limit, I'm assuming this was intentional. So it looks like most of the time was spent applying the update itself (https://github.com/kargakis/origin/blob/4f0dd02d4b63d33998441bc4d8c18b0076cff1ee/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/controller/controller_utils.go#L309), since we don't see an `UPDATE error:` line between the slow lines. This can mean a few different things:
1. You rc is starved of qps (if you just have one rc, a few nodes and services this shouldn't happen for 5s)
2. The rc manager is starved off cpu (what machine are you running the master on, and what else is on it?)
3. The apiserver is clogged 
4. Etcd is clogged 

if 3 or 4 is happening you shold see errors or weird silence in the respective logs

**Comment by smarterclayton** [comment]
I think in a few cases we've seen this while trying to scale down multiple rcs in succession, where it appears the RC manager gets hung up.  I believe in that env we had the rate limiter disabled - it looked like a starvation / exhaustion thing, but I don't have data to back that up.

**Comment by bprashanth** [comment]
#8822 should clear the water, after that I can add some clarity around when/if we hit rest client rate limits and we can take it from there

**Comment by kargakis** [comment]
@bprashanth the refactoring I've made in the loop doesn't change its behavior, it just makes more explicit what's happening. I'm running on an Intel Core i7-4600U CPU @ 2.10GHz, and run Kubernetes via OpenShift in a Vagrant environment. I'll run more tests and get back with verbose logging from the server.

**Comment by kargakis** [comment]
Hmm, the update time seems pretty random. Now after deleting 50 pods it just took 1s. 
```
I0527 15:50:34.929216   10575 controller_utils.go:306] Updating replica count for rc/ruby-hello-world-1, 50->0
I0527 15:50:35.936822   10575 controller_utils.go:311] Successfully updated rc/ruby-hello-world-1
```

@bprashanth generally I was wondering if we could update status.replica as the deletion of the pods happens?

**Comment by smarterclayton** [comment]
This sounds like an update is lost to a race somewhere, and as a result we have to wait for the sync interval / timeout in some cases.

----- Original Message -----

**Comment by bprashanth** [comment]
If the update is lost because of, say a CAS race, we requeue the rc so it should retry without waiting for resync (it's added at the tail of the queue for fairness): https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L389
When this happens you should also see the error in the logs.

> generally I was wondering if we could update status.replica as the deletion of the pods happens?

We could consider doing that. I guess you mean as each goroutine runs (and not when we spawn the goroutines). However on a decent network, the time between a goroutine sumitting a Delete rpc and the watch for the same arriving should be low. If each goroutine updated the status.Replicas it would probably exhaust qps and semi-dos the apiserver, so it's good to narrow down the cause as much as possible now.

**Comment by lavalamp** [comment]
> generally I was wondering if we could update status.replica as the deletion of the pods happens?

It should be updated when the deletion is observed, not when it's initiated. We want this field to follow reality, not precede it.

(imagine a client waiting for a RC stop to do an RC delete, and the RC manager sets replicas to 0 when it sends the deletions, and then the client sees this and deletes the RC, and then something goes wrong and the deletions fail-- this leaves stranded pods)

**Comment by bprashanth** [comment]
> We want this field to follow reality, not precede it.

If necessary one could assume that a 200 response from the Delete rpc is enough instead of waiting for the watch. I'll need a little convincing that this is necessary, though.  

>(imagine a client waiting for a RC stop to do an RC delete, and the RC manager sets replicas to 0 when it sends the deletions, and then the client sees this and deletes the RC, and then something goes wrong and the deletions fail-- this leaves stranded pods)

This is exactly what would happen if we updated the status.Replicas when we spawn the delete goroutines (i.e at the point you see the line "Too few replicas...")

**Comment by lavalamp** [comment]
> If necessary one could assume that a 200 response from the Delete rpc is enough instead of waiting for the watch. I'll need a little convincing that this is necessary, though.

I'd still prefer to decrement replicas after the deletion has actually propagated through the system. What I'd really like is to not decrement it until kubelet has actually deleted the pod, but that's not possible until we have some way of expressing the being-deleted stage in the API.

**Comment by bprashanth** [comment]
> What I'd really like is to not decrement it until kubelet has actually deleted the pod, but that's not possible until we have some way of expressing the being-deleted stage in the API.

Agreed. That would make stop even slower :)

**Comment by smarterclayton** [comment]
> On May 27, 2015, at 4:27 PM, Daniel Smith <notifications@github.com> wrote:
> 
> If necessary one could assume that a 200 response from the Delete rpc is enough instead of waiting for the watch. I'll need a little convincing that this is necessary, though.
> 
> I'd still prefer to decrement replicas after the deletion has actually propagated through the system. What I'd really like we is to not decrement it until kubelet has actually deleted the pod, but that's not possible until we have some way of expressing the being-deleted stage in the API.
> 
We have that with #8641.  But I don't think waiting for the pod to be fully deleted is the right behavior - I think you want to be creating new pods immediately after the deletion notice has been accepted by the API server, because in most cases you want additional capacity online immediately (many web frameworks will stop accepting new requests immediately after getting SIGTERM).

**Comment by davidopp** [comment]
I tend to agree with @lavalamp. Waiting is the conservative thing and arguably users are less likely to be negatively impacted if they are surprised that a new pod took a long time to spin up vs. surprised that they had more replicas running at the same time than they wanted (for example, they may not have enough quota to run more than the indicated number of pods or something like that). Maybe we should make it a pod-configurable option? 

The issue gets even more complicated with nominal services, sharded batch, and migration, where the service may just break if you spin up a replacement pod too soon, but you lose some of the pre-spinning benefit if you do it too late. I think we are likely to revisit this decision either way, but for now would vote for making it configurable or waiting for the deletion to propagate through the system.

**Comment by smarterclayton** [comment]
> On May 28, 2015, at 1:30 AM, David Oppenheimer <notifications@github.com> wrote:
> 
> I tend to agree with @lavalamp. Waiting is the conservative thing and arguably users are less likely to be negatively impacted if they are surprised that a new pod took a long time to spin up vs. surprised that they had more replicas running at the same time than they wanted (for example, they may not have enough quota to run more than the indicated number of pods or something like that). Maybe we should make it a pod-configurable option?
> 
> The issue gets even more complicated with nominal services, sharded batch, and migration, where the service may just break if you spin up a replacement pod too soon, but you lose some of the pre-spinning benefit if you do it too late.
> 
If you break because of pre-spin, you're already going to be broken because of the node controller behavior as soon as the node loses a connection to the master.  Are we doing users a favor by pretending that replication controller guarantees "no more than N instances will be running"?  And in the meantime, you reduce the availability of every service using a replication controller in the presence of failure.

This actually isn't a huge problem in most cases - I still think the issue here is a lost update, not the underlying semantics.  Observing the delete was recorded is sufficient (and with graceful in place, gives additional opportunities to observe deletion).

**Comment by bprashanth** [comment]
>This actually isn't a huge problem in most cases - I still think the issue here is a lost update, not the underlying semantics.

In the absense of unexpected update failures, an update to status.Replicas won't get lost. It might get delayed because of cpu or qps. 

The tradeoff with status.Replicas is that we want it to reflect the current state of the system, not some catch-up state. Eg:
```
t0: resize: 10->0
t1: the 0th update to status.replicas fails and the rc gets requeued
t1: resize 0->10
t2: the requed rc wakes up
  - the replicas are created 
  - the status.replica update fails again (if this passes we're safe)
  - the rc gets requeued
t3: the requeued rc wakes up 
  - status.replicas passes with replicas=10
```
In this scenario status.Replicas skips 0 and jumps to 10.

Unfortunately kubectl stop relies on status.Replicas, so if you resize an rc fast and the wrong updates fail, kubectl could take a long time even though the replicas are all dead. However:
* The pods are consistent since they went from 10->0->10
* Won't happen if the updates all pass
* We can engineer around it by having kubctl stop watch pods (https://github.com/GoogleCloudPlatform/kubernetes/issues/8572)

**Comment by lavalamp** [comment]
It's a client error to resize to 10 again if you're changing the pod in some way and you haven't observed the change to 0 succeed.

**Comment by bprashanth** [comment]
They could observe the success via listing pods though

**Comment by davidopp** [comment]
Decrementing RC.Status when Kubelet starts the grace period seems like the right thing to do. (I agree with Clayton that you can't design your app such that it breaks if there are more replicas running than you want, e.g. in the case where a node becomes partitioned from the master but not from other nodes so the NodeController marks the node dead and the RC creates a replacement pod but the node isn't dead and the original pod is still running.)  As @bprashanth pointed out if you really want to wait for successful deletion you can look at the PodStatuses.

But more generally, @bprashanth can you summarize what is left to do on this issue?

**Comment by bprashanth** [comment]
We should fix how kubectl stop rc interacts with the rc manager. The specific issue that causes confusion is as follows:

- We have an rc with 200 replicas, with 20qps this takes 20s to start
- Under normal conditions, we would create 200 replicas, and update status.Replicas=200 10s later
- If someone calls kubectl stop 1s after kubectl create, kubectl will observe the status.Replicas=0 and mistakenly assume it's work is done.

As @lavalamp argued, resizing an rc before having observed the expected size is a client error. But this is confusing, since kubectl returns immediately, so the user doesn't bother checking status.Replicas. This has always been the behavior of kubectl, the issue is exacerbated by the client qps.

There are a couple of solutions:
1. Kubectl watches pods instead of status.Replicas (@kargakis has a WIP up for this based of previous discussion https://github.com/GoogleCloudPlatform/kubernetes/pull/8999)
2. Always have kubectl stop specify a precondition of status.Replicas=spec.Replicas (kubectl stop already takes preconditions, I'm only taking about defaulting it)
3. Update status.Replicas from each goroutine, as we create replicas (instead of after the watch). This is  expensive for large rcs.

I'm for doing 1. @kargakis if you don't have time to finish up that pr I can take over.

**Comment by bgrant0607** [comment]
Kubectl should not be watching pods, for multiple reasons. First, I don't want to proliferate the logic for deciding which pods count and which ones don't. Also, we should be pushing more of the logic server-side, as discussed in #12143.

**Comment by bprashanth** [comment]
We already did https://github.com/GoogleCloudPlatform/kubernetes/issues/9147 to fix the problem described in my previous comment. We should push stop to the rc manage before optimizing its performance further.

**Comment by kargakis** [comment]
RC deletion is now an issue of the GC.
